src/python/pants/core/target_types.py

```diff
@@ -62,14 +63,100 @@
 from pants.util.meta import frozen_after_init
 from pants.util.strutil import softwrap
 
+# -----------------------------------------------------------------------------------------------
+# `per_platform` object
+# -----------------------------------------------------------------------------------------------
+_T = TypeVar("_T")
+
+
+@dataclass(frozen=True)
+class per_platform(Generic[_T]):
+    """An object containing differing homogenous platform-dependant values.
+
+    The values should be evaluated for the execution environment, and not the host environment
+    (I.e. it should be evaluated in a `rule` which requests `Platform`).
```

Comment:
This is more subtle, unfortunately. Rather than "any `@rule` which can get a `Platform`", this would instead need to be something like "not during dependency inference", essentially. https://github.com/pantsbuild/pants/issues/17354 covers trying to make that less error prone, but I'm not sure when that will happen.

Reply:
I think that nuance is captured in the `NOTE` paragraph later. I'd like to keep this first section high-level.